xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCRoomManager.java

```diff
@@ -285,11 +300,67 @@ void restoreCacheContentAfterJoin(@Nullable final Set<OccupantManager.Occupant>
                         for (OccupantManager.Occupant localOccupantToRestore : localOccupantsToRestore ) {
                             // Get the Role for the local occupant from the local representation of the room, and add that to the cluster room.
                             final MUCRole localOccupantRole = localRoom.getOccupantByFullJID(localOccupantToRestore.getRealJID());
+
                             if (localOccupantRole == null) {
                                 Log.trace("Trying to add occupant '{}' but no role for that occupant exists in the local room. Data inconsistency?", localOccupantToRestore.getRealJID());
                                 continue;
+                            } else {
+                                Log.trace("Found localOccupantRole {} for localOccupantToRestore {}, client route = {}", localOccupantRole, localOccupantToRestore.getRealJID(), XMPPServer.getInstance().getRoutingTable().getClientRoute(localOccupantToRestore.getRealJID()));
+                            }
+
+                            // OF-2165
+                            // Check if the nickname of this occupant already existed for another user in the room.
+                            // If it did, we need to kick the users out. With sincere apologies.
+                            String nickBeingAddedToRoom = localOccupantRole.getNickname();
+                            boolean occupantWasKicked = false;
+                            try {
+                                final List<MUCRole> existingOccupantsWithSameNick = roomInCluster.getOccupantsByNickname(nickBeingAddedToRoom);
+                                final List<JID> otherUsersWithSameNick = existingOccupantsWithSameNick.stream().map(MUCRole::getUserAddress).filter(bareJid -> !bareJid.equals(localOccupantRole.getUserAddress())).collect(Collectors.toList());
+                                Log.debug("Current cluster room occupants: {}", roomInCluster.getOccupants());
```

Comment:
There's a bit of surplus debug logging here that should probably be removed.

Reply:
Missed that one. Removed now.